common/flatpak-transaction-log.c

```diff
@@ -0,0 +1,287 @@
+ */
+FlatpakTransactionLog *
+flatpak_transaction_log_new (GFile *path)
+{
+  return g_object_new (FLATPAK_TYPE_TRANSACTION_LOG,
+                       "path", path,
+                       NULL);
+}
+
+static void
+flatpak_transaction_log_finalize (GObject *object)
+{
+  FlatpakTransactionLog *self = FLATPAK_TRANSACTION_LOG (object);
+
+  g_clear_object (&self->path);
+
+  G_OBJECT_CLASS (flatpak_transaction_log_parent_class)->finalize (object);
+}
+
+static void
+flatpak_transaction_log_get_property (GObject    *object,
+                                      guint       prop_id,
+                                      GValue     *value,
+                                      GParamSpec *pspec)
+{
+  FlatpakTransactionLog *self = FLATPAK_TRANSACTION_LOG (object);
+
+  switch (prop_id)
+    {
+    case PROP_PATH:
+      g_value_set_object (value, self->path);
+      break;
+
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+    }
+}
+
+static void
+flatpak_transaction_log_set_property (GObject      *object,
+                                      guint         prop_id,
+                                      const GValue *value,
+                                      GParamSpec   *pspec)
+{
+  FlatpakTransactionLog *self = FLATPAK_TRANSACTION_LOG (object);
+
+  switch (prop_id)
+    {
+    case PROP_PATH:
+      g_assert (self->path == NULL);
+      self->path = g_value_dup_object (value);
+      break;
+
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+    }
+}
+
+/* For efficiency's sake, we have a special "append only" log format here. This
+ * is a subset of JSON, but is not valid JSON in its own right - rather, each
+ * line is a new JSON object. Applications wishing to parse this file should
+ * read and parse each line independently. */
+static gboolean
+append_node_to_log_file (GFile        *file,
+                         JsonNode     *node,
+                         GCancellable *cancellable,
+                         GError       **error)
+{
+  g_autoptr(GFile) parent = g_file_get_parent (file);
+  g_autoptr(GFileOutputStream) stream = NULL;
+  g_autofree gchar *serialized = NULL;
+  gsize serialized_len = 0;
+  gsize written = 0;
+  g_autoptr(JsonGenerator) generator = NULL;
+  g_autoptr(GError) my_error = NULL;
+
+  if (!g_file_make_directory_with_parents (parent, cancellable, &my_error))
+    {
+      if (!g_error_matches (my_error, G_IO_ERROR, G_IO_ERROR_EXISTS))
+        {
+          g_propagate_error (error, g_steal_pointer (&my_error));
+          return FALSE;
+        }
+    }
+
+  generator = json_generator_new ();
+  json_generator_set_root (generator, node);
+  json_generator_set_pretty (generator, FALSE);
+
+  /* Resize the buffer to accomadate one extra newline and write that
+   * in place at serialized_len */
+  serialized = json_generator_to_data (generator, &serialized_len);
+  serialized = g_realloc (serialized, (serialized_len + 1) * sizeof (gchar));
```

Comment:
Instead of reallocating this, you could simply write the newline/terminator with a second call to `g_output_stream_write_all()`